### PR TITLE
Apps crypto free

### PR DESCRIPTION
**Pull request by richsalz**
This is based on #14193.  One corner case to apps/dgst to fix.

### CODE REVIEW

File: apps/dgst.c [hunk 1]
```diff
@@ -403,7 +403,7 @@ int dgst_main(int argc, char **argv)
     if (md == NULL) {
         EVP_MD_CTX *tctx;
         BIO_get_md_ctx(bmd, &tctx);
-        md = EVP_MD_CTX_md(tctx);
+        md = (EVP_MD *)EVP_MD_CTX_md(tctx);
```
Comment (levitte) [inline apps/dgst.c:406]: This change, or rather the `EVP_MD_free()` of this pointer down at the end, causes the following test error, which can now be observed in the CIs

```
free(): double free detected in tcache 2
../../util/wrap.pl ../../apps/openssl dgst -sign ../../test/testrsa.pem -out testrsa.sig ../../test/data.bin => 134
```

File: apps/dgst.c [hunk 2]
```diff
@@ -402,8 +402,9 @@ int dgst_main(int argc, char **argv)
 
     if (md == NULL) {
         EVP_MD_CTX *tctx;
+
         BIO_get_md_ctx(bmd, &tctx);
-        md = EVP_MD_CTX_md(tctx);
+        md = EVP_MD_CTX_get0_md(tctx);
```
Comment (levitte) [inline apps/dgst.c:407]: So here's a problem to keep track of: you must never try to free something you got from EVP_MD_CTX_get0_md()

Comment (levitte) [reply apps/dgst.c:407]: The same goes for an EVP_CIPHER returned by EVP_CIPHER_CTX_get0_cipher()

Comment (richsalz) [reply apps/dgst.c:407]: yes, this was raised before.  And I said I have thoughts on how to handle it.

Comment (levitte) [reply apps/dgst.c:407]: Well, there's an obvious solution: use the get1 function instead

**Comment by levitte** [comment]
It's been a while.  How goes the battle?

**Comment by richsalz** [comment]
Nice of you to check in with me :)  Was busy past two weeks on work and IETF virtual, picking things up now.
I have I think four bad-memory failure cases in `test_dgst` right now.  So, progress.

**Comment by levitte** [comment]
Please make sure to update this branch on top of #14193.  Call it a litmus test...

**Comment by richsalz** [comment]
This is new code, so it's not quite a litmus test. But yes, I will keep it more-or-less current with the work in the other PR.

**Comment by richsalz** [comment]
Thanks to @levitte contributions, in #14193 this now works. Still a DRAFT because that one needs to be merged first. If you look at this PR before, only look at the top commit.

File: apps/enc.c [hunk 3]
```diff
@@ -633,6 +633,7 @@ int enc_main(int argc, char **argv)
     BIO_free_all(out);
     BIO_free(benc);
     BIO_free(b64);
+    EVP_MD_free(dgst);
```
Comment (levitte) [inline apps/enc.c:636]: You should free `cipher` too

File: apps/lib/opt.c [hunk 4]
```diff
@@ -356,21 +356,25 @@ void print_format_error(int format, unsigned long flags)
...
 
 /* Parse a cipher name, put it in *EVP_CIPHER; return 0 on failure, else 1. */
-int opt_cipher(const char *name, const EVP_CIPHER **cipherp)
+int opt_cipher(const char *name, EVP_CIPHER **cipherp)
 {
-    *cipherp = EVP_get_cipherbyname(name);
+    *cipherp = (EVP_CIPHER *)EVP_get_cipherbyname(name);
     if (*cipherp != NULL)
         return 1;
+    *cipherp = EVP_CIPHER_fetch(NULL, name, NULL);
```
Comment (levitte) [inline apps/lib/opt.c:364]: You're not checking whether `*cipherp` is NULL, or most importantly, not.

Also, I think we concluded at some point that fetching should be done first, with legacy working as a fallback.

File: apps/lib/opt.c [hunk 5]
```diff
@@ -356,21 +356,25 @@ void print_format_error(int format, unsigned long flags)
...
-int opt_md(const char *name, const EVP_MD **mdp)
+int opt_md(const char *name, EVP_MD **mdp)
 {
-    *mdp = EVP_get_digestbyname(name);
+    *mdp = (EVP_MD *)EVP_get_digestbyname(name);
+    if (*mdp != NULL)
+        return 1;
+    *mdp = EVP_MD_fetch(NULL, name, NULL);
     if (*mdp != NULL)
         return 1;
```
Comment (levitte) [inline apps/lib/opt.c:379]: May I suggest somewhat less redundant code?

``` C
    *mdp = (EVP_MD *)EVP_get_digestbyname(name);
    if (*mdp == NULL)
       *mdp = EVP_MD_fetch(NULL, name, NULL);
    if (*mdp != NULL)
        return 1;
```

Also, just as with ciphers, perhaps fetch first, and legacy as fallback?

**Review by levitte — CHANGES_REQUESTED** [review]

File: apps/lib/opt.c (hunk 4 above)
Comment (richsalz) [reply apps/lib/opt.c:364]: Added NULL check, and fixed the order to be fetch then legacy.

File: apps/lib/opt.c (hunk 5 above)
Comment (richsalz) [reply apps/lib/opt.c:379]: I fixed the order to be fetch then legacy.
Saving one return statement, okay :)

Comment (levitte) [reply apps/lib/opt.c:380]: Even more terse (and more readable in my mind) would be:

``` C
    if ((*mdp = EVP_MD_fetch(NULL, name, NULL)) != NULL
        || (*mdp = (EVP_MD *)EVP_get_digestbyname(name)) != NULL)
        return 1;
```

Comment (richsalz) [reply apps/lib/opt.c:380]: in retrospect, I'm not going to do this.  The current code is clear and obvious and squeezing a handful of bytes or CPU cycles is the wrong trade-off.

**Comment by richsalz** [comment]
If I change `do_cipher` to do fetch before lookup I get many failures:
```
Test Summary Report
-------------------
20-test_dgst.t                   (Wstat: 256 Tests: 7 Failed: 1)
  Failed test:  6
  Non-zero exit status: 1
25-test_crl.t                    (Wstat: 256 Tests: 7 Failed: 1)
  Failed test:  5
  Non-zero exit status: 1
25-test_verify_store.t           (Wstat: 512 Tests: 10 Failed: 2)
  Failed tests:  1, 7
  Non-zero exit status: 2
25-test_req.t                    (Wstat: 256 Tests: 43 Failed: 1)
  Failed test:  15
  Non-zero exit status: 1
80-test_ca.t                     (Wstat: 256 Tests: 7 Failed: 2)
  Failed tests:  1, 7
  Non-zero exit status: 1
  Parse errors: Bad plan.  You planned 15 tests but ran 7.
80-test_tsa.t                    (Wstat: 256 Tests: 22 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
  Parse errors: Bad plan.  You planned 27 tests but ran 22.
90-test_store.t                  (Wstat: 512 Tests: 8 Failed: 0)
  Non-zero exit status: 2
  Parse errors: Bad plan.  You planned 426 tests but ran 8.
80-test_ssl_old.t                (Wstat: 2048 Tests: 11 Failed: 8)
  Failed tests:  1-8
  Non-zero exit status: 8
```

Many of them are because output changed when the FIPS provider is used first.  *WHY?*

For example in the CRL test:
```
# Got      SHA2-256 Fingerprint=B3:A9:FD:A7:2E:8C:3D:DF:D0:F1:C3:1A:96:60:B5:FD:B0:99:7C:7F:0E:E4:34:F5:DB:87:62:36:BC:F1:BC:1B
# Expected SHA256 Fingerprint=B3:A9:FD:A7:2E:8C:3D:DF:D0:F1:C3:1A:96:60:B5:FD:B0:99:7C:7F:0E:E4:34:F5:DB:87:62:36:BC:F1:BC:1B
not ok 5
```

One seems like a real error, found while running `test_verify_store`:
```
Writing new private key to 'keyCA.ss'
-----
malloc_consolidate(): invalid chunk size
../../util/wrap.pl ../../apps/openssl req -new -section userreq -config ../../test/ca-and-certs.cnf -out reqCA.ss -keyout keyCA.ss => 134
not ok 1 - make cert request
```

I can debug that, but need to know about the output differences first.

I will push a commit with the broken change.

**Comment by levitte** [comment]
Re the output difference, the provider implementations don't currently have a "main name", i.e. all names for one implementation are treated as equals. @t8m is working on a change to have a "canonical" name per implementation, which should "disappear" the problem.

It is interesting, though, that the changed order between fetching and getting legacy implementation should create such havoc...

**Comment by richsalz** [comment]
> It is interesting, though, that the changed order between fetching and getting legacy implementation should create such havoc... For the sake of this PR, I would suggest putting back the legacy first order.

I will do that.  Ping @t8m to consider taking a look at this issue when his work is done.

**Comment by richsalz** [comment]
Rebased onto master, removed WIP.

File: apps/ca.c [hunk 6]
```diff
@@ -795,7 +795,7 @@ int ca_main(int argc, char **argv)
      */
     if (def_ret == 2 && def_nid == NID_undef) {
         /* The signing algorithm requires there to be no digest */
-        dgst = EVP_md_null();
+        dgst = (EVP_MD *)EVP_md_null();
```
Comment (levitte) [inline apps/ca.c:798]: This proliferation of casts has me wonder if we should un-constify the likes of `EVP_md_null()`, `EVP_get_digestbyname()`, ...

Comment (levitte) [reply apps/ca.c:798]: (the cast could be done inside those functions)

Comment (richsalz) [reply apps/ca.c:798]: I would not make `EVP_md_null` etc non-const. It is a transition for moving to a pure-fetch environment.

Comment (levitte) [reply apps/ca.c:798]: Yeahok...  I can't help but find it jarring, but you do have a point.

**Review by levitte — APPROVED** [review]

**Comment by levitte** [comment]
@openssl, second review needed

**Review by beldmit — APPROVED** [review]
LGTM

File: apps/ca.c (hunk 6 above)
Comment (t8m) [reply apps/ca.c:798]: Unconstifying would break existing callers so that is another reasons not to do that.

**Pull request closed without merging**